### PR TITLE
Fix: safe call to F1 MLC scores

### LINKED ISSUE #57: Tracked multilabel metrics are wrong
The bug is here https://github.com/autrainer/autrainer/blob/2cdd8bb64ca6550a69a58e22bf291664c6d2a9f0/autrainer/training/training.py#L814

and is caused by the use of `average='micro'` or `average='macro'` with `sklearn` when computed for each label separately. I have not yet figured out exactly why this happens (i.e. what goes on internally) but results are very inflated compared to the "normal" computation of the metric when using all target columns (these are the results that are printed).

This has potentially impacted the results reported in https://huggingface.co/autrainer/edansa-2019-cnn10-32k-t

### PR DESCRIPTION
**Pull request by ATriantafyllopoulos**
This closes #57 by introducing a `_safe_f1_score` function that checks for the dimensionality of predictions/labels and overrides the `average` argument if necessary..

When testing this with `ToyTabular-MLC` this resulted in the results being stored in `test_holistic.yaml` being identical to the ones printed at the end of training.

Nevertheless, I want to add some proper tests before merging

### CODE REVIEW

**Comment by ramppdev** [comment]
> Nevertheless, I want to add some proper tests before merging

This is a good idea, the question is if we want to create a separate function for stratified evaluation to the abstract metric.
This function could default to the normal `__call__` as before for all metrics and to per-class / -target evaluation in the case of multi-label classification / multi-target regression.
Then the functionality could be clearer for extensibility.

Right now all F1 scores would just produce the same per-class score right?
To still respect the F1 averaging strategy, we should then use the normal `__call__` function also for stratified evaluation, i think this would still report wrong scores now?
https://github.com/autrainer/autrainer/blob/2cdd8bb64ca6550a69a58e22bf291664c6d2a9f0/autrainer/training/training.py#L825-L826

**Comment by ATriantafyllopoulos** [comment]
> This is a good idea, the question is if we want to create a separate function for stratified evaluation to the abstract metric.

You mean that there would be a function with the following prototype?

```python
def stratified(self, groundtruth: pd.DataFrame, predictions: pd.DataFrame, columns: List[str]):
```

> Right now all F1 scores would just produce the same per-class score right?

Yes, you are right. This brings us back to the problem that some of those metrics don't make sense the way we do them. For instance, `micro` averaging would need to compute the false positives etc. over all classes. This cannot be substituted by averaging. So maybe your proposal is better.

**Comment by ATriantafyllopoulos** [comment]
Updated:

1. Use `tracker.predictions` and `tracker.targets` to simplify evaluations

2. Compute metrics globally for storing into `all`. So this now should be "correct", and is not a simple average of metrics per label. Therefore, the per-label `f1 scores` for `micro` and `macro` will be the same (corresponding to `average='binary'`, but will be set to the correct value for `all`

3. Cast `tracker.predictions` to `np.array` (at some point, this had been turned to a list by most/all target transforms)

**Comment by ATriantafyllopoulos** [comment]
30d76ec added some very simple checks for the f1-score, which cover the new functionality as well

File: autrainer/training/training.py [hunk 1]
```diff
@@ -800,34 +800,20 @@ def _disaggregated_evaluation(self, df, groundtruth, stratify):
...
-                total /= len(self.data.target_column)
-                results[metric.name]["all"] = total
-            else:
-                results[metric.name]["all"] = metric(
-                    groundtruth[self.data.target_column], df["predictions"]
-                )
+                    results[metric.name][col] = metric(
+                        tracker.targets[:, idx],
+                        tracker.predictions[:, idx],
+                    )
```
Comment (ramppdev) [inline autrainer/training/training.py:813]: > > This is a good idea, the question is if we want to create a separate function for stratified evaluation to the abstract metric.
> 
> You mean that there would be a function with the following prototype?
> 
> ```python
> def stratified(self, groundtruth: pd.DataFrame, predictions: pd.DataFrame, columns: List[str]):
> ```

Would you prefer this? I mean it works now nicely, but if a user wants to add a new multi-class / multi-target metric, they would need to perform the same checks as in `_safe_f1_score`. If we just call `metric.stratified` here it could be more clear what happens?

File: autrainer/training/training.py [hunk 2]
```diff
@@ -800,34 +800,20 @@ def _disaggregated_evaluation(self, df, groundtruth, stratify):
         TODO: Need to define schema in the docs.
 
         """
-        df = df.set_index(groundtruth.index)
```
[the remark this answers is not in the document]
Comment (ATriantafyllopoulos) [reply autrainer/training/training.py:803]: Ah, good point. This is not something I had in mind, I thought `_disaggregated_evaluations` is only called for dev/test (i.e. without shuffling). 

However, I am not using `df` at all here, but rather get the indices from `groundtruth`:

https://github.com/autrainer/autrainer/blob/30d76ecf5e6071ec864fa1b54266663ec26290dd/autrainer/training/training.py#L824-L828

This is still wrong because `idx` corresponds to the original index. I can change this to:

```python
idx = groundtruth.loc[groundtruth[s] == v].index
                    # Map groundtruth indices to tracker indices
                    # This accounts for random shuffling
                    indices = [
                        i
                        for i, x in enumerate(tracker.indices)
                        if x in idx
                    ]
                    results[metric.name][v] = metric(
                        tracker.targets[indices],
                        tracker.predictions[indices],
                    )
``` 
Do you see a faster implementation?

See 16f06dc

P.S. This also needs a test though..

File: autrainer/training/training.py (hunk 1 above)
Comment (ATriantafyllopoulos) [reply autrainer/training/training.py:813]: Yes, I think this is cleaner. I will switch to that.

Comment (ATriantafyllopoulos) [reply autrainer/training/training.py]: Changed

File: autrainer/training/training.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ATriantafyllopoulos) [reply autrainer/training/training.py:803]: I moved `disaggregated_evaluation` to `autrainer.training.utils` and so was able to test it in `test_metrics`, I've also changed the interface a bit

File: autrainer/training/utils.py [hunk 3]
```diff
@@ -1,9 +1,70 @@
...
+            # multi-target regression
+            # loops over all targets and computes the metric for them
+            for idx, col in enumerate(target_column):
+                results[metric.name][col] = metric.unitary(
+                    targets[:, idx],
+                    predictions[:, idx],
+                )
+        results[metric.name]["all"] = metric(targets, predictions)
+        for s in stratify:
+            if isinstance(target_column, list):
```
[the remark this answers is not in the document]
Comment (ramppdev) [reply autrainer/training/utils.py:56]: Wow yes this is a big mistake i made in #53! I changed `assert` to `if` without inverting the condition ...

File: autrainer/training/training.py (hunk 1 above)
Comment (ramppdev) [reply autrainer/training/training.py]: Yes, this looks great and is clean!

File: autrainer/metrics/multi_label_classification.py [hunk 4]
```diff
@@ -18,6 +19,25 @@ def __init__(self):
             average="macro",
         )
 
+    def unitary(self, y_true: np.ndarray, y_pred: np.ndarray) -> float:
+        """Unitary evaluation of metric.
+
+        Metric computed with `average='binary'`
+        i.e. only accounting for positive label.
```
Comment (ramppdev) [inline autrainer/metrics/multi_label_classification.py:26]: for `the` positive label / for positive label`s`. Or leave as is :D

File: autrainer/training/training.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ramppdev) [reply autrainer/training/training.py:803]: Okay yeah i think this should be fine and the tests (last one for disaggregated evaluation) also cover shuffled indices right?

**Review by ramppdev — APPROVED** [review]
I would say the PR looks great, thank you very much!

File: autrainer/training/training.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ATriantafyllopoulos) [reply autrainer/training/training.py:803]: Not all tests cover disaggregated indices, only the last one
